ACG provisioning platform API spec

rlupton20-msft opened the pull request:
# Data Plane API - Pull Request
<!-- 🚨🚨🚨 Replace this line with a summary and reason for these changes to your data plane API 🚨🚨🚨 -->

## API Info: The Basics
Most of the information about your service should be captured in the issue that serves as your [*API Spec engagement record*](https://aka.ms/azsdk/onboarding/restapischedule).

* Link to API Spec engagement record issue: https://github.com/Azure/azure-rest-api-specs/issues/25899

Is this review for (select one):

- [ ] a private preview
- [ ] a public preview
- [x] GA release

### Change Scope

<sup>This section will help us focus on the specific parts of your API that are new or have been modified. <br/>Please share a link to the design document for the new APIs, a link to the previous API Spec document (if applicable), and the root paths that have been updated. </sup>
* Design Document: can provide details on request.
* Previous API Spec Doc:
* Updated paths:

### Viewing API changes

For convenient view of the API changes made by this PR, refer to the URLs provided in the table 
in the `Generated ApiView` comment added to this PR. You can use ApiView to show API versions diff. 

### Suppressing failures

If one or multiple validation error/warning suppression(s) is detected in your PR, please follow the 
[Swagger-Suppression-Process](https://dev.azure.com/azure-sdk/internal/_wiki/wikis/internal.wiki/85/Swagger-Suppression-Process) 
to get approval.

## ❔Got questions? Need additional info?? We are here to help!

<details>
  <summary> Contact us!</summary>

The [Azure API Review Board](https://aka.ms/azsdk/onboarding/restapischedule) is dedicated to helping you create amazing APIs. You can read about our mission and learn more about our process on our [wiki](https://aka.ms/azsdk/onboarding/restapischedule).
* 💬 [Teams Channel](https://teams.microsoft.com/l/channel/19%3a3ebb18fded0e47938f998e196a52952f%40thread.tacv2/General?groupId=1a10b50c-e870-4fe0-8483-bf5542a8d2d8&tenantId=72f988bf-86f1-41af-91ab-2d7cd011db47)
* 💌 [email](mailto://azureapirbcore@microsoft.com)

</details>

<details>
  <summary>Click here for links to tools, specs, guidelines & other good stuff</summary>

### Tooling

 * [Open API validation tools](https://aka.ms/swaggertools) were run on this PR. Go here to see [how to fix errors](https://aka.ms/ci-fix)
 * [Spectral Linting](https://aka.ms/style)
 * [Open API Hub](https://aka.ms/openapiportal)

### Guidelines & Specifications

 * [Azure REST API Guidelines](https://aka.ms/guidelines)
 * [OpenAPI Style Guidelines](https://aka.ms/style)
 * [Azure Breaking Change Policy](https://aka.ms/AzBreakingChangesPolicy)

### Helpful Links

 * [Azure DevTools Wiki](https://aka.ms/azapi)

</details>

Commit:
ACG provisioning platform API spec

weshaggard commented:
@lmazuel any idea who would be the best reviewer for this?

rlupton20-msft commented:
[README.md](https://github.com/Azure/azure-rest-api-specs/files/12870186/README.md)

@tg-msft @mikekistler I've attached some docs detailing how we expected users to connect and use this API. This should cover the core concepts (accounts and numbers), and give a bunch of likely usage scenarios (with responses etc).

rlupton20-msft commented on specification/voiceservices/data-plane/Microsoft.VoiceServices/stable/v1.0/swagger.json:
```diff
@@ -0,0 +1,1052 @@
+      "type": "oauth2",
+      "description": "Azure Active Directory OAuth2 Flow.",
+      "flow": "implicit",
+      "authorizationUrl": "https://login.microsoftonline.com/common/oauth2/authorize",
```
This is the flow that we need: https://learn.microsoft.com/en-us/azure/active-directory/develop/v2-oauth2-client-creds-grant-flow#get-a-token but it's swagger isn't compatible with tenant specific token URLs.

Please advise how we should handle this.

rlupton20-msft replied:
Believe this is now resolved.

mikekistler reviewed (changes requested):
See [comments in review issue](https://github.com/Azure/azure-rest-api-specs/issues/25899#issuecomment-1768975450)

mikekistler commented on specification/voiceservices/data-plane/Microsoft.VoiceServices/stable/2023-10-01/swagger.json:
```diff
@@ -0,0 +1,1015 @@
+  "paths": {
+    "/account/{accountName}/numbers": {
+      "get": {
+        "operationId": "Account_GetNumbers",
```
This operation and several others are missing the required "api-version" query parameter.
```
>spectral lint $f | grep api-version
  32:22      warning  az-version-policy                Operation does not define an "api-version" query parameter.                paths./account/{accountName}/numbers.get.parameters
  98:22      warning  az-version-policy                Operation does not define an "api-version" query parameter.                paths./account/{accountName}/numbers/batch.post.parameters
 154:22      warning  az-version-policy                Operation does not define an "api-version" query parameter.                paths./account/{accountName}/numbers/{phoneNumber}.delete.parameters
 208:22      warning  az-version-policy                Operation does not define an "api-version" query parameter.                paths./account/{accountName}/numbers/{phoneNumber}.get.parameters
 268:22      warning  az-version-policy                Operation does not define an "api-version" query parameter.                paths./account/{accountName}/numbers/{phoneNumber}.put.parameters
 344:22      warning  az-version-policy                Operation does not define an "api-version" query parameter.                paths./accounts.get.parameters
 397:22      warning  az-version-policy                Operation does not define an "api-version" query parameter.                paths./accounts/{accountName}.delete.parameters
 441:22      warning  az-version-policy                Operation does not define an "api-version" query parameter.                paths./accounts/{accountName}.get.parameters
 499:22      warning  az-version-policy                Operation does not define an "api-version" query parameter.                paths./accounts/{accountName}.put.parameters
 565:22      warning  az-version-policy                Operation does not define an "api-version" query parameter.                paths./numbers.get.parameters
```

rlupton20-msft replied:
Ah, apologies - I'd created the parameter description but not included it on the endpoints - fixed.

mikekistler commented on specification/voiceservices/data-plane/Microsoft.VoiceServices/stable/2023-10-01/swagger.json:
```diff
@@ -0,0 +1,1015 @@
+            "description": "An offset into the collection of the first record to be returned."
+          }
+        ],
+        "responses": {
```
Operations with a pageable response should specify the `x-ms-pageable` extension.

mikekistler commented on specification/voiceservices/data-plane/Microsoft.VoiceServices/stable/2023-10-01/swagger.json:
```diff
@@ -0,0 +1,1015 @@
+          "200": {
+            "description": "Success",
+            "schema": {
+              "$ref": "#/definitions/NumberResponseDtoPaginatedListDto"
```
Schema names should be clear and concise.
```suggestion
              "$ref": "#/definitions/NumberList"
```

mikekistler commented on specification/voiceservices/data-plane/Microsoft.VoiceServices/stable/2023-10-01/swagger.json:
```diff
@@ -0,0 +1,1015 @@
+            "type": "string",
+            "description": "Name of the account. Can only contain letters, numbers, underscores and dashes. Can be up to 100 characters in length."
+          },
+          {
+            "in": "body",
+            "name": "body",
+            "schema": {
+              "$ref": "#/definitions/BatchNumbersPutDto"
+            }
+          }
```
Should this body parameter be marked as required?

rlupton20-msft replied:
Yes - went over all of the requests with bodies and fixed this.

mikekistler commented on specification/voiceservices/data-plane/Microsoft.VoiceServices/stable/2023-10-01/swagger.json:
```diff
@@ -0,0 +1,1015 @@
+        "description": "Gets the details of a number assigned to an account, including the services for which the number is enabled."
+      },
+      "put": {
+        "operationId": "Number_Put",
```
The operationId should be descriptive of the operation and not the underlying HTTP method used.
```suggestion
        "operationId": "Number_CreateOrReplace",
```

rlupton20-msft replied:
Done. Also checked the other operationsIds (apart from those flagged, the batch endpoint also needed some tweaking).

mikekistler commented on specification/voiceservices/data-plane/Microsoft.VoiceServices/stable/2023-10-01/swagger.json:
```diff
@@ -0,0 +1,1015 @@
+            "type": "string",
+            "description": "Phone number in E.164 format."
+          },
+          {
+            "in": "header",
+            "name": "If-Match",
+            "type": "string"
+          },
+          {
+            "in": "header",
+            "name": "If-None-Match",
+            "type": "string"
+          },
```
All parameters should have a description for the benefit of the generated documentation.

rlupton20-msft replied:
Added documentation.

mikekistler commented on specification/voiceservices/data-plane/Microsoft.VoiceServices/stable/2023-10-01/swagger.json:
```diff
@@ -0,0 +1,1015 @@
+            "name": "If-None-Match",
+            "type": "string"
+          },
+          {
+            "in": "body",
+            "name": "body",
+            "schema": {
+              "$ref": "#/definitions/NumberDetailsDto"
+            }
+          }
```
Should this body parameter be marked required?

mikekistler commented on specification/voiceservices/data-plane/Microsoft.VoiceServices/stable/2023-10-01/swagger.json:
```diff
@@ -0,0 +1,1015 @@
+        "description": "Gets the details of the specified account. An account represents a customer."
+      },
+      "put": {
+        "operationId": "Account_Put",
```
The operationId should be descriptive of the operation and not the underlying HTTP method used.
```suggestion
        "operationId": "Account_CreateOrReplace",
```

rlupton20-msft replied:
Updated. Checked other operationIds too.

mikekistler commented on specification/voiceservices/data-plane/Microsoft.VoiceServices/stable/2023-10-01/swagger.json:
```diff
@@ -0,0 +1,1015 @@
+      "properties": {
+        "etag": {
+          "type": "string",
+          "x-nullable": true,
```
Avoid the use of `x-nullable`.

rlupton20-msft replied:
Removed.

mikekistler reviewed (changes requested):
I flagged a few more issues to resolve.

rlupton20-msft commented:
Addressed the new issues, and took over the lint output and fixed up things in the same vein (and other things I thought worthwhile). Thanks for reviewing! Back over to you.

mikekistler reviewed (changes requested):
Just a few more minor issues to resolve.

rlupton20-msft commented on specification/voiceservices/data-plane/Microsoft.VoiceServices/stable/2023-10-01/swagger.json:
```diff
@@ -0,0 +1,1063 @@
+      "x-ms-enum": {
+        "name": "SubdomainProvisioningStatus",
+        "modelAsString": true
+      }
```
Done.

rlupton20-msft commented on specification/voiceservices/data-plane/Microsoft.VoiceServices/stable/2023-10-01/swagger.json:
```diff
@@ -0,0 +1,1063 @@
+        "additionalProperties": {
+          "type": "object",
+          "description": "Additional properties that can be provided about the error."
+        }
```
Done.

rlupton20-msft commented on specification/voiceservices/data-plane/Microsoft.VoiceServices/stable/2023-10-01/swagger.json:
```diff
@@ -0,0 +1,1063 @@
+            "description": "New details of the number to create or replace."
+          }
+        ],
+        "responses": {
```
Done.

rlupton20-msft commented on specification/voiceservices/data-plane/Microsoft.VoiceServices/stable/2023-10-01/swagger.json:
```diff
@@ -0,0 +1,1063 @@
+    "AccountsList": {
+      "additionalProperties": false,
+      "properties": {
+        "records": {
```
Changed to `value`.

rlupton20-msft commented on specification/voiceservices/data-plane/Microsoft.VoiceServices/stable/2023-10-01/swagger.json:
```diff
@@ -0,0 +1,1093 @@
+      "type": "oauth2",
+      "description": "Azure Active Directory OAuth2 Flow.",
+      "flow": "application",
+      "tokenUrl": "https://login.microsoftonline.com/common/oauth2/v2.0/token",
```
I believe our service uses a tenant specific authentication endpoint for this flow. How do we express that here?

rlupton20-msft replied:
Our flow looks like this, https://learn.microsoft.com/en-us/entra/identity-platform/v2-oauth2-client-creds-grant-flow#get-a-token but as far as I can tell CI rejects any templating, and I've been unable to find any docs on how this is meant to be done.

mikekistler replied:
I think the SDK generators all have ways to provide more specific information on authentication, so what you have here should be fine.

rlupton20-msft replied:
Thanks - that makes sense.

mikekistler reviewed (approved):
Looks good! 👍 

Thanks for the fixes!

mikekistler commented:
Not sure why this PR has no assignee -- it looks like none was ever assigned.

@tjprescott this PR is ready to merge.
